tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c

```diff
@@ -0,0 +1,171 @@
+	returned_keys_refs[params_it - 1] = returned_key;
+
+	zassert_true(returned_key != NULL, "bt_keys_get_addr() returned a non-valid reference");
+	zassert_true(!bt_addr_le_cmp(&returned_key->addr, &addr),
+		     "bt_keys_get_addr() set incorrect address %s value, expected %s",
+		     bt_addr_le_str(&returned_key->addr), bt_addr_le_str(&addr));
+	zassert_false(check_key_pool_is_empty(),
+		      "List is empty, (ID, Address) pair insertion failed");
+}
+
+/*
+ *  Test no equal references returned by bt_keys_get_addr()
+ *
+ *  Constraints:
+ *   - Keys pool has been filled
+ *
+ *  Expected behaviour:
+ *   - All returned references are different from each other
+ */
+void test_no_equal_references(void)
+{
+	struct bt_keys *keys_pool = bt_keys_get_key_pool();
+
+	for (uint32_t i = 0; i < ARRAY_SIZE(returned_keys_refs); i++) {
+		struct bt_keys *returned_ref = returned_keys_refs[i];
+
+		zassert_equal_ptr(keys_pool + i, returned_ref,
+				  "bt_keys_get_addr() returned unexpected reference at slot %u", i);
+	}
+}
+
+/*
+ *  Test getting a valid key reference by a matching ID and address
+ *
+ *  Constraints:
+ *   - ID and address pair has been inserted in the list
+ *
+ *  Expected behaviour:
+ *   - A valid reference is returned by bt_keys_get_addr() that
+ *     matches the one returned after adding the ID and address pair
+ *   - ID value matches the one passed to bt_keys_get_addr()
+ *   - Address value matches the one passed to bt_keys_get_addr()
+ */
+void test_get_key_by_matched_id_and_address(void)
+{
+	bt_addr_le_t addr;
+	struct bt_keys *returned_key, *expected_key_ref;
+	uint8_t id = current_params_vector->id;
+
+	bt_addr_le_copy(&addr, current_params_vector->addr);
+
+	returned_key = bt_keys_get_addr(id, &addr);
+	/* 'params_it' is incremented before running the test case */
+	expected_key_ref = returned_keys_refs[params_it - 1];
+
+	zassert_true(returned_key != NULL, "bt_keys_get_addr() returned a non-valid reference");
+	zassert_equal_ptr(returned_key, expected_key_ref,
+			  "bt_keys_get_addr() returned unexpected reference");
+}
+
+/* Initialize test variables */
+static void test_round_initialization(void)
+{
+	params_it = 0;
+	current_params_vector = NULL;
+}
+
+/* Setup test variables */
+static void unit_test_setup(void)
+{
+	zassert_true((params_it < (ARRAY_SIZE(testing_id_addr_pair_lut))),
+		     "Invalid testing parameters index %u", params_it);
+	current_params_vector = &testing_id_addr_pair_lut[params_it];
+	params_it++;
+}
+
+void test_main(void)
+{
+	/* This routine should setup the conditions for subsequent tests
+	 *   - Fully fill keys pool
+	 *   - Store returned references
+	 */
+	ztest_test_suite(bt_keys_get_addr_get_non_existing_keys,
+		ztest_unit_test(test_round_initialization),
+		ztest_unit_test(test_list_is_empty_at_startup),
+		LISTIFY(CONFIG_BT_MAX_PAIRED, REGISTER_SETUP_TEARDOWN, (,), test_fill_keys_pool),
+		ztest_unit_test(test_no_equal_references));
```

Comment:
I think you got your impression that these are just a setup functions because the same functions and names are used in other functions UT projects like `bt_keys_foreach_bond` and `bt_keys_foreach_type`.

I can agree with your proposal and call these functions directly without considering them test cases, except for this one in `bt_keys_get_addr/src/main.c`.

This one is the only real one that is testing `test_get_non_existing_keys` and may be the name `test_fill_keys_pool` is a bit misleading and gives a wrong impression.

So, I agree to apply your suggestions in other projects because they are actually used as a setup functions.
Here, we can just re-name `test_fill_keys_pool` to `test_get_non_existing_keys` because this is its main job and filling the keys pool can be considered as a side-effect of it.

Reply:
so I think I got half of my point over (which is nice). But I still think you shouldn't be relying on the side effects of previous test cases: each individual UT should be runnable by itself in isolation and you should be able to shuffle them without encountering issues.

Reply:
RFC, I think we need to settle on the definition of `UT` in your comment.

In my opinion, It should be used to refer to a `test suite` which is composed of one or more test functions.
Running the `test suite` verifies a certain behavior and it may be required to run test functions in a certain order.

I think `test suites` should be independent from each other and their order shouldn't affect the result.
But a test function inside a `test suite` can be dependent on the output from the previous one.

Reply:
I guess that can work, if you abuse the `suite` to be the UT and the `tests` inside it (except one) to be setup-teardown functions.

In that case, could you rename each `test_` fn that is actually setting up the state for the real logic test `setup_`? same for `teardown_` fns. That way, the reader (which could be us in a few years) will understand what's happening where, and know not to reorder the functions not prefixed by `test_`.

More consensus on order-dependent tests being a bad practice:
https://stackoverflow.com/a/53383608
https://softwareengineering.stackexchange.com/a/368803
top of that comment https://softwareengineering.stackexchange.com/a/335433

Reply:
> In that case, could you rename each `test_` fn that is actually setting up the state for the real logic test `setup_`?

Unfortunately, that isn't possible, all test functions that will run using the framework must starts wit `test_`.
This can be found here https://docs.zephyrproject.org/3.1.0/develop/test/ztest.html

You can only setup a single setup function to be called before running the test function.

Anyway, I extracted the setup functions outside the `test suite` and they are now explicitly called before running the `test suite` as you previously suggested